### PR TITLE
Chip: component enhancement

### LINKED ISSUE #1191: Chip: implement version 2
Implement version 2 of the Chip component in accordance with its design in Figma
https://www.figma.com/file/zPvPwPgb4EsYFXKC2nhzSG/Infineon-DDS-%7C-Main?node-id=17762%3A11219&mode=dev

Requirements: 
- max two selected options are displayed
- the number indicator appears when more than 2 options are selected
- clicking the x-button unselects all options

### PR DESCRIPTION
**Pull request by akashyeole**
By creating this pull request you agree to the terms in CONTRIBUTING.md.
https://github.com/Infineon/.github/blob/master/CONTRIBUTING.md
--- DO NOT DELETE ANYTHING ABOVE THIS LINE ---

Description
- Implemented `<ifx-chip-item>` which replaces the use of dropdown menus.
- Implemented the new variant of chip -- multiselect. 
- Component redesign and other changes. 
- Updated the storybook controls for Chip.

Related Issue
#1191 

<details>
  <summary>📦 Published PR as canary version: <code>25.0.0--canary.1441.82a12c4f68670317b6c7608e024bd100c3702d71.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @infineon/infineon-design-system-stencil@25.0.0--canary.1441.82a12c4f68670317b6c7608e024bd100c3702d71.0
  # or 
  yarn add @infineon/infineon-design-system-stencil@25.0.0--canary.1441.82a12c4f68670317b6c7608e024bd100c3702d71.0
  ```
</details>

### CODE REVIEW

File: packages/components/src/components/chip/chip-item/chip-item.tsx [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+@Component({
+    tag: 'ifx-chip-item',
+    styleUrl: 'chip-item.scss',
+    shadow: true
+})
+
+export class ChipItem {
+    /**
+     * Reference to this component.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip-item/chip-item.tsx:19]: would also remove all these comments above the variable names, as the variable names should be self explanatory

File: packages/components/src/components/chip/chip-item/chip-item.tsx [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    /**
+     * The prop allows to set the initial *selected* status of the Chip Item.
+     * 
+     * @Default `false`
+     */
+    @Prop({ mutable: true, reflect: true }) selected: boolean = false;
+
+    /**
+     * Listenting for ifxChipItem event to unselect the previously selected Item.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip-item/chip-item.tsx:48]: this comment also isnt needed as the function naming tells already

File: packages/components/src/components/chip/chip-item/chip-item.tsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            if (this.chipState.emitIfxChipItem){
+                this.emitIfxChipItemEvent();
+            } else {
+                this.chipState.emitIfxChipItem = true;
+            }
+        }
+    } 
+    
+    /**
+     * A function extracts and returns the Item label.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip-item/chip-item.tsx:78]: comment not need either, as getItemLabel is self explanatory

File: packages/components/src/components/chip/chip-item/chip-item.tsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+     * 
+     * @returns string
+     */
+    getItemLabel(): string {
+        return this.chipItem.innerText as string;
+    }
+
+    /**
+     * Toggles the selection of the Chip Item and emits the ifxChipItem event.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip-item/chip-item.tsx:87]: same here

File: packages/components/src/components/chip/chip-item/chip-item.tsx [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    render() {
+        return (
+            <div class={`chip-item chip-item--${this.chipState.size} 
+                         chip-item--${(this.selected && this.chipState.variant) === 'single' ? 'selected' : ''}`} 
+                tabIndex={0}
+                onClick={() => {this.handleItemClick()}}
+                onKeyDown={(e) => {this.handleItemKeyDown(e)}}>
+
+                {/* Checkbox; renders only in 'multi' variant. */}
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip-item/chip-item.tsx:138]: can you explain the reason for why you use checkbox + label here and what was the advantage over using multiselect for example?

File: packages/components/src/components/chip/chip.tsx [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        const path = event.composedPath();
+        const chipWrapper: HTMLElement = this.chip.shadowRoot.querySelector('.chip__wrapper');
+        const chipDropdown: HTMLElement = this.chip.shadowRoot.querySelector('.chip__dropdown');
+        if (!path.includes(chipDropdown) && !path.includes(chipWrapper) && this.opened) {
+            this.toggleDropdownMenu();
+        }
+    }
+
+    /**
+     * Listening to ifxChipItem event to update the selected option(s).
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip.tsx:80]: remove comment

File: packages/components/src/components/chip/chip.tsx [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+     */
+    @State() opened: boolean = false;
+
+    /**
+     * State to store the selected options of the chip component.
+     */
+    @State() selectedOptions: Array<ChipItemEvent> = [];
+
+    /**
+     * Closing the dropdown menu when clicked outside of it.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip.tsx:67]: remove comment

File: packages/components/src/components/chip/chip.tsx [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+        /* Emitting ifxChange with the selected options. */
+        if (eventDetail.emitIfxChange){
+            this.ifxChange.emit({ previousSelection: previousSelection, 
+                                  currentSelection: this.selectedOptions});
+        }
+    }
+    
+    /**
+     * Returns the list of reference to the Chip Items.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip.tsx:125]: this function name also pretty well explains what it does and thus, we dont need the comment

File: packages/components/src/components/chip/chip.tsx [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+     * 
+     * @returns `NodeList<HTMLIfxChipItemElement>`
+     */
+    getChipItems(): NodeList {
+        return this.chip.querySelectorAll('ifx-chip-item');
+    }
+
+    /**
+     * Returns the selected option label in *single* variant.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip.tsx:134]: same here

File: packages/components/src/components/chip/chip.tsx [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+                    continue;
+                }
+                break;
+            }
+            return optionsLabel;
+        }   
+    }
+
+    /**
+     * Opens or Closes the dropdown menu.
```
Comment (verena-ifx) [inline packages/components/src/components/chip/chip.tsx:162]: and here

**Review by verena-ifx — CHANGES_REQUESTED** [review]

**Comment by akashyeole** [comment]
> The way the chip component's dropdown (the checkbox+label class) is implemented, it gets hidden behind other components, when opened.

Which other components for example?
I tried placing button, badge and accordion below chip and dropdown is always on the top these components.

**Comment by verena-ifx** [comment]
One more comment @akashyeole could you rename the custom event that is emitted to ifxChipItemSelect or something like that?

**Comment by verena-ifx** [comment]
> > The way the chip component's dropdown (the checkbox+label class) is implemented, it gets hidden behind other components, when opened.
> 
> Which other components for example? I tried placing button, badge and accordion below chip and dropdown is always on the top these components.

I used it inside the advanced table, but let me clarify with the designers first if you should even be able to open it, as your filters are in the sidebar and thats where you select. in theory :D
image

**Comment by verena-ifx** [comment]
> @verena-ifx 'Deploy canary releases and update example apps on...' check is still failing after removing node-sass from dependencies. I don't it has anything to do with node-sass.

when i checked before i saw the working directory wasnt clean. This normally gets solved by a clean install though

**Comment by verena-ifx** [comment]
@akashyeole I just discussed this regarding the design when used within the table. I think it makes sense to add an additional property "readonly". When set to true, the dropdown menu wont open and it will only show the selected items (which are selected from the table filter)
But let me get some further information first :)

**Comment by tishoyanchev** [comment]
@akashyeole 
- Rename event name ifxChipChange to ifxChange for consistency

**Comment by akashyeole** [comment]
> @akashyeole
> 
> * Rename event name ifxChipChange to ifxChange for consistency

It was 'ifxChange' before, Verena changed it as she mentioned it in the comment.
But I also think that it should be 'ifxChange'.  @verena-ifx thoughts?

**Comment by verena-ifx** [comment]
> > @akashyeole
> > 
> > * Rename event name ifxChipChange to ifxChange for consistency
> 
> It was 'ifxChange' before, Verena changed it as she mentioned it in the comment. But I also think that it should be 'ifxChange'. @verena-ifx thoughts?

mhh for select and multiselect we have ifxSelect. So its not ifxChange for all the components right now. Which benefits do you see in having it ifxchange? I think if we wanna be more unanimous, we should go through all the components and classify them. But before that maybe look at other design systems?

File: packages/components/src/components/chip/chip.tsx [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+          {
+            !this.readOnly && (this.variant !== 'multi' || (this.variant === 'multi' && this.selectedOptions.length === 0)) &&
+            <div class='wrapper__open-button'>
+              <ifx-icon key={1} icon={`chevrondown16`} />
+            </div>
+          }
+  
+          {
+            ((this.selectedOptions.length >= 1) && this.variant === 'multi') &&
+            <div class='wrapper__unselect-button' onClick={(e) => { this.handleUnselectButtonClick(e) }}>
```
Comment (akashyeole) [inline packages/components/src/components/chip/chip.tsx:218]: @verena-ifx can you add readOnly check for onClick on this line?

**Comment by tishoyanchev** [comment]
@verena-ifx 
- Can you rename the event name ifxChipChange to ifxChange for consistency?
- The number indicator showing the remaining selected options is not working correctly. I have 4 items. When I select 3, all 3 are shown + the number indicator is showing "+1", indicating that there is one more selected item that is not shown despite that not being the case. If I select all 4, all 4 are displayed, and the number indicator is showing "+2", despite there not being 2 more options selected. All selected options are always displayed, but after the 3rd one, they're also shown in the number indicator. Pic related: 
chip issue 1
- Cannot run application in the example apps. 
React error: 
react error
Chip not included in Vue and Vanilla-cdn

**Comment by akashyeole** [comment]
> @verena-ifx
> 
> * Can you rename the event name ifxChipChange to ifxChange for consistency?
> * The number indicator showing the remaining selected options is not working correctly. I have 4 items. When I select 3, all 3 are shown + the number indicator is showing "+1", indicating that there is one more selected item that is not shown despite that not being the case. If I select all 4, all 4 are displayed, and the number indicator is showing "+2", despite there not being 2 more options selected. All selected options are always displayed, but after the 3rd one, they're also shown in the number indicator. Pic related:
> 
> chip issue 1 - Cannot run application in the example apps. React error: react error Chip not included in Vue and Vanilla-cdn

@verena-ifx 

In chip.tsx, on line:198, you should use the this.getSelectedOptions() function. This function handles the joining of selected options and returns at max 2 selected options.

File: packages/components/src/components/chip/chip.tsx (hunk 11 above)
Comment (akashyeole) [inline packages/components/src/components/chip/chip.tsx:218]: @verena-ifx 
Bug: Selected options get unselected when `readOnly` is set and cross button is clicked in multi variant.

https://github.com/user-attachments/assets/00ab1da6-123d-4773-ae89-dc7fb2699ca0

Please readOnly prop check for onClick on line 218.

**Comment by verena-ifx** [comment]
> > @verena-ifx
> > 
> > * Can you rename the event name ifxChipChange to ifxChange for consistency?
> > * The number indicator showing the remaining selected options is not working correctly. I have 4 items. When I select 3, all 3 are shown + the number indicator is showing "+1", indicating that there is one more selected item that is not shown despite that not being the case. If I select all 4, all 4 are displayed, and the number indicator is showing "+2", despite there not being 2 more options selected. All selected options are always displayed, but after the 3rd one, they're also shown in the number indicator. Pic related:
> > 
> > chip issue 1 - Cannot run application in the example apps. React error: react error Chip not included in Vue and Vanilla-cdn
> 
> @verena-ifx
> 
> In chip.tsx, on line:198, you should use the this.getSelectedOptions() function. This function handles the joining of selected options and returns at max 2 selected options.

regarding the event naming, i thnk that should be a separate issue, because we dont have consistent event naming right now. this PR is already too big cause of the outdated stencil output target dependencies :/

**Comment by tishoyanchev** [comment]
@akashyeole, @verena-ifx 
- If placeholder is not set, the colons remains. Pic related: 
chip placeholder
- In storybook, read-only page, clicking on the X will unselect all options
-suggestion: when read-only is true, remove the "x" button rather than making it disabled

**Pull request merged**